text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+- Feature Name: `never_patterns`
+- Start Date: 2024-10-27
+- RFC PR: [rust-lang/rfcs#3719](https://github.com/rust-lang/rfcs/pull/3719)
+- Rust Issue: [rust-lang/rust#118155](https://github.com/rust-lang/rust/issues/118155)
+
+# Summary
+[summary]: #summary
+
+A `!` pattern indicates a type with no valid values. It is used to indicate an impossible case when matching an empty type in unsafe code.
+
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+Note: this RFC is purely about a new pattern syntax. It does not propose changes to exhaustiveness checking or to the operational semantics of uninhabited types.
+
+
+# Motivation
+[motivation]: #motivation
+
+Rust's unsafe semantics are access-based: the validity of data only matters when reading from/writing to a place. When pattern-matching, patterns only access the data (discriminants/values) they need to choose the right arm.
+
+Empty types are funny because the natural way of matching them is to write no arms at all:
+
+```rust
+fn empty_tup<T>(tup: (T, !)) -> ! {
+    match tup {}
+}
+```
+
+Here the absence of an arm plays the role of a "read discriminant" kind of operation. This is fine in this case, but around unsafe code that interacts with possibly-uninitialized data, accesses should be explicitly visible in the pattern.
+
+Today, when matching empty types inside places that may contain uninitialized data, rust requires a dummy match arm:
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup {
+            _ => unreachable_unchecked(),
+            // or
+            (_, x) => match x {}
+        }
+    }
+}
+
+union MyUnion<T: Copy> {
+    uninit: (),
+    value: (T, !),
+}
+impl<T: Copy> MyUnion<T> {
+    unsafe fn assume_init(self) -> ! {
+        match self {
+            MyUnion { value: (_, x) } => x,
+        }
+    }
+}
+```
+
+This RFC proposes a new `!` pattern that works as an explicit "access data" or "assert validity" operation for uninhabited types. The examples above become:
```

Comment:
I think this line is the key point.  It defines what the programmer means when they write `!`.

However, in safe code (eg, when the scrutinee is a safe reference), this fact - "this is actually sane" - is something that should always be true and shouldn't need to be explained by the programmer.

ISTM that this is also a pedagogical problem.  If this is needed in safe Rust, auithors of that safe Rust will need an explanationi what this thing is for and what it means, but we definitley don't want to burden them thoughts about validity/safety invariants etc.  Nor do we want to suiggest to them in discussion of `match` that values might not be valid.  But there is no way to explain this `!` without supposing that `&T` might not reference a proper `T`.  Authors of safe Rust are entitled to assume that `&T` always does reference a valid `T`.

Therefore, `!` should not be required in safe Rust.  What does that mean precisely?  One possible answer:
 * `!` is not required when dereferencing a safe reference, field, etc. (recursively)
 * Except, within an `unsafe` block, we emit a lint saying "something something uninhabited types validity something something requirement something"

Reply:
I don't think we want to make syntax depend on whether something is syntactically inside or outside an unsafe block. That kind of context-dependence can become quite surprising.

So if we do anything like this, IMO it should be fully based on the type of the scrutinee (including the types of its subexpressions, e.g. whether a union field projection is involved).

Reply:
But just to check, would you consider a possible path forward looking something like:

* Never patterns match types with a certain property, independent of location (this property could be "safety-based uninhabitness").
* When creating a `match` statement on a type with a different property, the compiler inserts a never pattern at the bottom of the match, independent of location (this property could be, e.g. "doesn't contain pointers or union field projections or partially uninhabited tuples" - roughly as [per the auto-never blog post](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/))).
* We add a lint for when the auto-never might cause issues - which could be dependent on location?

I think such a lint being location-sensitive would be justifiable in that it doesn't change syntax, but just adds extra checks? Or at least Niko was happy with that in 2018.

Reply:
FWIW the path forward I intend to propose is simply to allow omitting arms for `&!` and similar types (more generally, if the type is safely uninhabited). My understanding is that making unsafe references is a massive footgun anyway (because of e.g. auto(de)ref).

Unsafe code authors that do that anyway can write never patterns for such cases (they will not be marked as unreachable), and we may add a lint that requires them if users find this useful.

I have moved away from explaining things in terms of "auto-never" out of personal preference, but you could explain this by saying that we always auto-never except inside pointers and unions.

> Never patterns match types with a certain property, independent of location

This is indeed precisely the intent of this here RFC.

Reply:
Given `ptr: *const &!`, I don't think `match *ptr {}` should ever build. That should require `match *ptr { ! }`.

IOW, this should depend on whether the matched-on place is a safe place or not.

Reply:
Indeed, I include this in the exception that "we don't auto-never inside pointers". By "inside pointers" I mean in terms of places.

Reply:
> Given ptr: *const &!, I don't think match *ptr {} should ever build. That should require match *ptr { ! }.
> 
> IOW, this should depend on whether the matched-on place is a safe place or not.

Okay, understood - so in your view auto-never behaviour (if/when we have it or an equivalent) would need to vary be place, and shouldn't just be a static property of the matched type.

But never patterns, as syntax, should be place invariant. This makes sense.

Reply:
If I may now stop discussion of auto-never and similar considerations: I'd like to keep the discussions on-topic of the never pattern syntax proposal, and discuss changes to which matches are accepted etc to some other place. For example https://github.com/rust-lang/rust/issues/131452